torchrl/modules/models/model_based_models.py

```diff
@@ -0,0 +1,316 @@
+        loc, scale = self.backbone(state, belief)
+        return loc, scale
+
+
+class ObsEncoder(nn.Module):
+    """Observation encoder network.
+
+    Takes an pixel observation and encodes it into a latent space.
+
+    Reference: https://arxiv.org/abs/1803.10122
+
+    Args:
+        depth (int): Number of hidden units in the first layer.
+    """
+
+    def __init__(self, depth=32):
+        super().__init__()
+        self.encoder = nn.Sequential(
+            nn.LazyConv2d(depth, 4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(depth, depth * 2, 4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(depth * 2, depth * 4, 4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(depth * 4, depth * 8, 4, stride=2),
+            nn.ReLU(),
+        )
+
+    def forward(self, observation):
+        *batch_sizes, C, H, W = observation.shape
+        if len(batch_sizes) == 0:
+            end_dim = 0
+        else:
+            end_dim = len(batch_sizes) - 1
+        observation = torch.flatten(observation, start_dim=0, end_dim=end_dim)
+        obs_encoded = self.encoder(observation)
+        latent = obs_encoded.reshape(*batch_sizes, -1)
+        return latent
+
+
+class ObsDecoder(nn.Module):
+    """Observation decoder network.
+
+    Takes the deterministic state and the stochastic belief and decodes it into a pixel observation.
+
+    Reference: https://arxiv.org/abs/1803.10122
+
+    Args:
+        depth (int): Number of hidden units in the last layer.
+    """
+
+    def __init__(self, depth=32):
+        super().__init__()
+        self.state_to_latent = nn.Sequential(
+            nn.LazyLinear(depth * 8 * 2 * 2),
+            nn.ReLU(),
+        )
+        self.decoder = nn.Sequential(
+            nn.LazyConvTranspose2d(depth * 4, 5, stride=2),
+            nn.ReLU(),
+            nn.ConvTranspose2d(depth * 4, depth * 2, 5, stride=2),
+            nn.ReLU(),
+            nn.ConvTranspose2d(depth * 2, depth, 6, stride=2),
+            nn.ReLU(),
+            nn.ConvTranspose2d(depth, 3, 6, stride=2),
+        )
+        self._depth = depth
+
+    def forward(self, state, rnn_hidden):
+        latent = self.state_to_latent(torch.cat([state, rnn_hidden], dim=-1))
+        *batch_sizes, D = latent.shape
+        latent = latent.view(-1, D, 1, 1)
+        obs_decoded = self.decoder(latent)
+        _, C, H, W = obs_decoded.shape
+        obs_decoded = obs_decoded.view(*batch_sizes, C, H, W)
+        return obs_decoded
+
+
+class RSSMRollout(nn.Module):
+    """Rollout the RSSM network.
+
+    Given a set of encoded observations and actions, this function will rollout the RSSM network to compute all the intermediate
+    states and beliefs.
+    Here, we use the previous posterior as the prior for the next time step. At the first time step, we use the an empty prior to start the rollout.
+    At the end, we end up with all the intermediate states and beliefs.
```

Comment:
Let's reformulate this a bit no? Like "The forward method returns a stack of all intermediate states and believes" or something like it?